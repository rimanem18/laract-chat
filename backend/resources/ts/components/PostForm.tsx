import React from 'react'
import {
  useAppDispatch,
  useParamGroupId,
  usePostContent,
  useUserId,
} from '../app/hooks'
import { Link, useParams } from 'react-router-dom'
import { getContent, postMessage } from '../slices/PostSlice'

const PostForm = () => {
  const dispatch = useAppDispatch()
  const userId = useUserId()
  const postContent = usePostContent()
  const groupId = useParamGroupId()

  if (groupId === undefined) {
    return null
  }

  // 投稿
  const postMessageHandler = (
    e:
      | React.KeyboardEvent<HTMLTextAreaElement>
      | React.FormEvent<HTMLFormElement>
  ) => {
    if (postContent === '') return
    e.preventDefault()
    dispatch(
      postMessage({
        userId: userId,
        groupId: Number(groupId),
        content: postContent,
      })
    )
  }

  const changeContentHandler = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    dispatch(getContent(e.target.value))
  }

  const onSubmitKeyUpHandler = (
    e: React.KeyboardEvent<HTMLTextAreaElement>
  ) => {
    // ctrl+Enter で送信
    if (e.code === 'Enter' && e.ctrlKey) {
      postMessageHandler(e)
    }
  }

  return (
    <>
      {userId === 0 ? (
        <p>
          書き込みをするには <Link to="/login">ログイン</Link> してください。{' '}
        </p>
      ) : (
        <form className="form post-form mt-1" onSubmit={postMessageHandler}>
          <textarea
            data-testid="textarea"
            className="post-form__input"
            name="content"
            value={postContent}
            onChange={changeContentHandler}
            onKeyUp={onSubmitKeyUpHandler}
            autoFocus
          ></textarea>
          <button className="btn btn-primary mx-1 align-top" type="button">
            投稿
          </button>
        </form>
      )}
    </>
  )
}

export default React.memo(PostForm)
