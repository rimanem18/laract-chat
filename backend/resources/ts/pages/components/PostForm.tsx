import React, { useState } from 'react'
import axios from 'axios'
import { useAppDispatch, useAppSelector } from '../../app/hooks'
import { selectChatMessages } from '../../features/ChatMessagesSlice'
import { selectUser } from '../../features/UsersSlice'

const PostForm = () => {
  const [userId, setUserId] = useState(1)
  const [content, setContent] = useState('')
  const dispatch = useAppDispatch()
  const chatMessages = useAppSelector(selectChatMessages)
  const user = useAppSelector(selectUser)

  // 投稿
  const postMessage = async (e: any) => {
    e.preventDefault()

    axios
      .post('/api/chat_messages/post', {
        userId: user.id,
        content: content,
      })
      .then((res) => {
        // 成功したらメッセージクリア
        setContent('')
        console.log(res.data)
      })
      .catch((err) => {
        console.log(err)
      })
  }

  return (
    <>
      <form onSubmit={postMessage}>
        <div>
          <p>{user.id}</p>
          <p>{user.name}</p>
          <p>{user.email}</p>
        </div>
        <textarea
          name="content"
          value={content}
          onChange={(e) => setContent(e.target.value)}
        ></textarea>
        <button type="submit">投稿</button>
      </form>
    </>
  )
}

export default PostForm
