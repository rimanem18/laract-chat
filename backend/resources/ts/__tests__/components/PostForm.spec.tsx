import React from 'react'
import '@testing-library/jest-dom'
import { fireEvent, render } from '@testing-library/react'
import PostForm from '../../components/PostForm'

// Hooks の Mock
const mockUseAppDispatch = jest.fn()
const mockUseAppSelector = jest.fn()
jest.mock('../../app/hooks', () => ({
  useAppDispatch:
    () =>
    (...args: any[]) =>
      mockUseAppDispatch(...args),
  useAppSelector:
    () =>
    (...args: any[]) =>
      mockUseAppSelector(...args),
  useUserId: () => useUserIdMock(),
  usePostContent: () => usePostContentMock(),
}))

const mockValues = {
  userId: 1,
  content: 'Hello',
}
const useUserIdMock = jest.fn().mockReturnValue(mockValues.userId)
const usePostContentMock = jest.fn().mockReturnValue(mockValues.content)

// Setup
const setup = () => {
  const screen = render(<PostForm />)
  const textarea = screen.getByTestId('textarea') as HTMLTextAreaElement
  return {
    textarea,
    ...screen,
  }
}

// テスト開始
describe('PostForm', () => {
  it('入力値と postSlice content の文字列が同じ', () => {
    const { textarea } = setup()

    expect(textarea.value).toBe(mockValues.content)
  })

  it('onChange が発火しても文字列は同じ', () => {
    const { textarea } = setup()

    fireEvent.change(textarea, {
      target: {
        value: 'Hello World!',
      },
    })
    expect(textarea.value).toBe(mockValues.content)
  })
})
