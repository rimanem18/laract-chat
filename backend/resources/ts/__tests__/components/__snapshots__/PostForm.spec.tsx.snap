// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostForm snapshot 1`] = `
<form
  className="post-form"
  onSubmit={[Function]}
>
  <div
    className="MuiFormControl-root MuiTextField-root post-form__input css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    onKeyUp={[Function]}
  >
    <div
      className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={true}
        className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        data-testid="post-form"
        disabled={false}
        name="content"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value="Hello"
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-1a1ihhj"
        >
          <span
            className="notranslate"
            dangerouslySetInnerHTML={
              Object {
                "__html": "&#8203;",
              }
            }
          />
        </legend>
      </fieldset>
    </div>
  </div>
  <button
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1kuq5xv-MuiButtonBase-root-MuiIconButton-root"
    disabled={false}
    onBlur={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="SendIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M2.01 21 23 12 2.01 3 2 10l15 2-15 2z"
      />
    </svg>
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</form>
`;
